### PR TITLE
Improve github action workflow

### PR DESCRIPTION
**Pull request by nyurik**
P.S. Note that CI currently fails - not certain why yet

I copied some of these from many of my crates like https://github.com/nyurik/sqlite-hashes (eventually I think it is better to use `just` to simplify the process, and make it identical for the developer and CI)

* removed the `no-libvarnish` job -- not used
* remove rust installation - it is always pre-installed unless you want an older version, e.g. for msrv test
* use build cache
* add all sorts of tests
* no need to do a loop over vmod - enough to just say `cargo ... --workspace`
* not much value in `RUST_BACKTRACE=1` I think - if it fails, longer backtrace won't tell you much in CI
* add doc testing

### CODE REVIEW

**Comment by gquintard** [comment]
don't worry about the CI failure, it did it's job and quote a bug :) I'll fix it (in short: we can be called with 0-length buffer, with a NULL pointer, and `std::slice::from_raw_parts_mut()` is rightfully unhappy with it)

re: `RUST_BACKTRACE=1`, I actually just added it, and it allowed be to debug the issue above, so I would keep it :-D

TIL `cargo ... --workspace` that's great, and much simpler

I'd like to keep `no-libvarnish`, it makes sure that crates.io doesn't blow up when creating the docs without `varnish` being installed

**Comment by nyurik** [comment]
thx for the explanation, updated

File: .github/workflows/tests.yaml [hunk 1]
```diff
@@ -1,36 +1,50 @@
 name: Build
-on: [push]
+
+on:
```
Comment (gquintard) [inline .github/workflows/tests.yaml:3]: any reason to limit ourselves here?

File: .github/workflows/tests.yaml [hunk 2]
```diff
@@ -1,36 +1,50 @@
...
+      - name: install varnish-dev
+        run: |
+          curl -s https://packagecloud.io/install/repositories/varnishcache/varnish75/script.deb.sh | sudo bash
+          sudo apt-get install -y varnish-dev
+      - name: test-fmt
+        run: cargo fmt --all -- --check
+      - name: check
+        run: RUSTFLAGS='-D warnings' cargo check --workspace --all-targets
+      - name: build vmod
+        run: cargo build --all-targets
```
Comment (gquintard) [inline .github/workflows/tests.yaml:32]: you need to build all the workspace as the `cargo test` below will require the vmods to be built

File: .github/workflows/tests.yaml (hunk 1 above)
Comment (nyurik) [reply .github/workflows/tests.yaml:3]: this is a common pattern for FOSS projects:
* only build the "release" branches (like `main`)
* when you or someone else pushes modification branches, there is little value in building those branches directly. Instead, you want to build the pull request **into** one of the release branches (main)
* this allows you and everyone else contributing to this project to work on the same system - me or you or anyone must create a PR for it to be built

File: .github/workflows/tests.yaml (hunk 2 above)
Comment (nyurik) [reply .github/workflows/tests.yaml:32]: i was a bit confused by that - test (in theory) must build the code. How is it possible for the test to run without building it first?

File: .github/workflows/tests.yaml (hunk 1 above)
Comment (gquintard) [reply .github/workflows/tests.yaml:3]: I mean, I have no objections, but we could also leave it as a free for all, it's not like we have millions of commits every day :-)

File: .github/workflows/tests.yaml (hunk 2 above)
Comment (nyurik) [reply .github/workflows/tests.yaml:32]: hmm, i guess in theory `test` might skip the step in which build places final binaries to the `release|debug` dir... Ok, will have to simply modify check -> build.

File: .github/workflows/tests.yaml (hunk 1 above)
Comment (nyurik) [reply .github/workflows/tests.yaml:3]: ??  it is free for all - my point is that you don't want to compile it 3 times:  on branch creation, on pull creation, and on merge to `main` -- simply because the first two are identical, and you should have PR before merging.

Comment (nyurik) [reply .github/workflows/tests.yaml:3]: P.S. note that if you modify the branch from which a PR was created, that triggers rebuild on both `branch` and `pull` -- so you get two builds on each change

Comment (gquintard) [reply .github/workflows/tests.yaml:3]: from my understanding, currently we run CI for every commit (i.e. free for all), and with this PR we'll only run for a subset of commits

The proposed way works for me, we're good

Comment (nyurik) [reply .github/workflows/tests.yaml:3]: not exactly - it runs when **you** commit anything to your repo.

* When I create a new PR, I first create a branch -- which trigger a local action run  (on my own fork). The results are local to my own fork.
* Afterwards, I create a PR, but you do NOT see the results of a build -- because the build happened on BRANCH before the PR was created (i.e. the build is not associated with the PR).
* Whenever I update the branch (e.g. per your feedback), the build once again happens locally on my fork - the PR does not show its results

So in short, it is far better to build on PR change rather than branch change - because it keeps all those build results in one place - associated with the PR

Comment (gquintard) [reply .github/workflows/tests.yaml:3]: ooooooh, gotcha, you're right, it's much better

**Comment by gquintard** [comment]
ok to merge for me, shall I press the green button?

**Comment by nyurik** [comment]
yep :)

**Comment by nyurik** [comment]
btw, you may want to do "squash only merges" - keeps history cleaner

**Comment by gquintard** [comment]
I prefer to just rebase, merging makes it harder to `git bisect`*, feel free to squash and rebase though. I'll merge later today to give you some time if you want to do that

* yes, the history is a mess a the moment, the pro/con of working alone on a code base, but hey, that's the intention that counts!

**Comment by nyurik** [comment]
its already on top, feel free to merge whenever, i won't change
